File: 30. Doubly linked list.py
class Node:
    def __init__(self,data):
        self.data=data; self.next=None; self.prev=None

class DoublyLinkedList:
    def __init__(self): self.head=None
    def push_front(self,val):
        n=Node(val); n.next=self.head
        if self.head: self.head.prev=n
        self.head=n
    def insert_after(self,prev,val):
        if not prev: return
        n=Node(val); n.next=prev.next; prev.next=n; n.prev=prev
        if n.next: n.next.prev=n
    def delete_node(self,del_node):
        if not self.head or not del_node: return
        if self.head==del_node: self.head=del_node.next
        if del_node.next: del_node.next.prev=del_node.prev
        if del_node.prev: del_node.prev.next=del_node.next
    def print_forward(self):
        cur=self.head
        while cur:
            print(cur.data, end=' '); cur=cur.next
        print()

if __name__=='__main__':
    dll=DoublyLinkedList(); dll.push_front(2); dll.push_front(1); dll.insert_after(dll.head,3); dll.print_forward(); dll.delete_node(dll.head.next); dll.print_forward()

File: 30. Doubly linked list.cpp
#include <iostream>
using namespace std; struct Node{int data; Node* next; Node* prev; Node(int d):data(d),next(nullptr),prev(nullptr){}
};
void push_front(Node*& head,int val){ Node* n=new Node(val); n->next=head; if(head) head->prev=n; head=n; }
void insert_after(Node* prev,int val){ if(!prev) return; Node* n=new Node(val); n->next=prev->next; prev->next=n; n->prev=prev; if(n->next) n->next->prev=n; }
void delete_node(Node*& head, Node* del){ if(!head || !del) return; if(head==del) head=del->next; if(del->next) del->next->prev=del->prev; if(del->prev) del->prev->next=del->next; delete del; }
void print_forward(Node* head){ while(head){ cout<<head->data<<" "; head=head->next; } cout<<endl; }
int main(){ Node* head=nullptr; push_front(head,2); push_front(head,1); insert_after(head,3); print_forward(head); delete_node(head, head->next); print_forward(head); return 0; }
